pallets/rewards/src/tests/fuzzy.rs

```diff
@@ -0,0 +1,138 @@
+	Unstake((DomainId, CurrencyId), Account, Balance),
+	Claim((DomainId, CurrencyId), Account),
+	Distribute(Vec<Group>, Balance),
+}
+
+struct TestState<Rewards> {
+	total_distributed: Balance,
+	accounts_used: BTreeSet<Account>,
+	_reward_system_t: std::marker::PhantomData<Rewards>,
+}
+
+impl<Rewards> Default for TestState<Rewards> {
+	fn default() -> Self {
+		Self {
+			total_distributed: 0,
+			accounts_used: BTreeSet::default(),
+			_reward_system_t: Default::default(),
+		}
+	}
+}
+
+impl<Rewards> TestState<Rewards>
+where
+	Rewards: DistributedRewards<GroupId = u32, Balance = Balance>
+		+ AccountRewards<Account, Balance = Balance, CurrencyId = (DomainId, CurrencyId)>,
+{
+	fn apply_command(&mut self, command: Command) -> DispatchResult {
+		match command {
+			Command::AttachCurrency(dom_curr, group) => {
+				Rewards1::attach_currency(dom_curr, group)?;
+			}
+			Command::Stake(dom_curr, account, amount) => {
+				Rewards::deposit_stake(dom_curr, &account, amount)?;
+				self.accounts_used.insert(account);
+			}
+			Command::Unstake(dom_curr, account, amount) => {
+				Rewards::withdraw_stake(dom_curr, &account, amount)?;
+				self.accounts_used.insert(account);
+			}
+			Command::Claim(dom_curr, account) => {
+				Rewards::claim_reward(dom_curr, &account)?;
+				self.accounts_used.insert(account);
+			}
+			Command::Distribute(groups, reward) => {
+				self.total_distributed += Rewards::distribute_reward(reward, groups)?
+					.iter()
+					.filter_map(|e| e.ok())
+					.fold(0, |acc, group_reward| acc + group_reward);
+			}
+		};
+
+		Ok(())
+	}
+
+	/// Checks the system invariant. Valid for `base` and `gap` mechanisms
+	fn validate(&mut self) {
+		let total_claimed = self.accounts_used.iter().fold(0, |acc, account| {
+			acc + free_balance(CurrencyId::Reward, account)
+		});
+
+		assert_eq!(self.total_distributed, total_claimed + rewards_account());
+	}
+}
+
+fn evaluate_sample<Rewards>(commands: impl IntoIterator<Item = Command>)
+where
+	Rewards: DistributedRewards<GroupId = u32, Balance = Balance>
+		+ AccountRewards<Account, Balance = Balance, CurrencyId = (DomainId, CurrencyId)>,
+{
+	new_test_ext().execute_with(|| {
+		let mut state = TestState::<Rewards>::default();
+
+		for command in commands {
+			// We do not care if we fail applying the command.
+			// We only care if the invariant is preserved even if we fail doing it.
+			state.apply_command(command.clone()).ok();
+		}
+
+		state.validate();
+	});
+}
+
+mod fuzzy {
+	use super::*;
+
+	/// A fuzzer that only generates and tests one hardcoded command combination
+	/// from the next matrix:
+	///
+	/// | Action     | Participants | Calls  | Total  |
+	/// |-           |-             |-       |-       |
+	/// | stake      | A, B         | 2      | 4      |
+	/// | unstake    | A, B         | 2      | 4      |
+	/// | claim      | A, B         | 2      | 4      |
+	/// | distribute |  -           | 2      | 2      |
+	///
+	/// It uses 1 group, 1 domain and 1 currency.
```

Comment:
These could be added to the matrix too, and it should work without extra modifications in the test code.